crates/bevy_app/src/preferences.rs

```diff
@@ -0,0 +1,275 @@
+use bevy_ecs::system::Resource;
+use bevy_reflect::{Map, Reflect, TypePath};
+
+use crate::Plugin;
+
+/// Adds application [`Preferences`] functionality.
+pub struct PreferencesPlugin;
+
+impl Plugin for PreferencesPlugin {
```

Comment:
From my perspective, Plugins already have a preferences API:

```rust
pub struct AssetPlugin {
    pub file_path: String,
    pub processed_file_path: String,
    pub watch_for_changes_override: Option<bool>,
    pub mode: AssetMode,
}
```

Notably, these preferences are stored on Apps (and accessible) at runtime:

```rust
let asset_plugin = app.get_added_plugins::<AssetPlugin>()[0];
```

Not the _ideal_ runtime API (not accessible on World, returns a Vec), but thats a solvable problem.

Also, notably, the vast majority of these plugin settings could be serializable.

This is the current "plugin settings" API, so any proposal should describe how it relates to it / justify its existence relative to it. What does `Preferences` enable that the current plugin settings approach could not support / evolve into? Can (and should) we unify the two concepts? Should we _remove_ plugin settings as they are currently defined?

Note that we moved away from "plugin settings as resources" in [this PR](https://github.com/bevyengine/bevy/pull/6336) in favor of the current approach. This `Preferences` proposal has a lot in common with that pattern (just using a centralized resource instead of individual resources). Can you read through the rationale in that PR for moving away and give your thoughts?

Reply:
The (unclear) distinction here is that using `Preferences` gives you a place to declare state that you, as the plugin author want to be globally persistent.

> Notably, these preferences are stored on Apps (and accessible) at runtime:

I had no idea this was the case, and I don't know if the ecosystem does either. I was under the impression that plugin state was set on init, and never updatable at runtime. Glad to be made aware of that, but I suppose that highlights a problem if I didn't even know this was a blessed pattern.

Stepping back from the current implementation questions, the user need here is for some idiomatic, easy-to-use place to declare some plugin state is persistent and _global_ to the application. This is distinct from, say, serializing a game save or project file, which is not global.

From that point of view, `Plugin` is a place to hold in memory state tied to a specific plugin, but there is no distinction or affordances to declare that some part of that state is global and (generally-speaking) hot reloadable. Additionally, there is no way to turn all of that global persistent state into a single, serializable structure, e.g. `settings.ron` or `preference.json`. I think that's one of the hardest things to accomplish: gathering all global state into a single place, turning it into trait soup, and being able to serde it to a dynamic format is not easy for most users.

To me, the benefit of something like `Preferences`, is that I can have my in memory representation of all plugin settings, like

```rs
pub struct AssetPlugin {
    pub file_path: String,
    pub processed_file_path: String,
    pub watch_for_changes_override: Option<bool>,
    pub mode: AssetMode,
}
```

and additionally, I can declare my public-facing preferences, which might look completely different, and require extra processing or runtime steps to apply to my application - or even writing back to the preferences file because a setting is invalid:

```rs
pub enum AssetPreferences {
    Default,
    TurboMode,
    Custom(/* ... /*)
}
```
 If I squint, it also seems trivial to add callbacks, so I could update my plugin settings when my preferences entry changes, or vice versa.

> Can you read through the rationale in that PR for moving away and give your thoughts?

I don't think this conflicts with that, but it is at least clear that the distinction needs to be made much more obvious to developers. 
- "I have state specific to a plugin": great, add it as part of your plugin struct
- "I want some of that state to persist in a global preferences file": define that, and add it to the `Preferences`. 

If this approach is acceptable, we might want to integrate this more closely with plugins, so that adding a preferences entry requires implementing a trait that allows conversion to and from the preference and plugin types. Something vaguely along the lines of:

```rs
pub trait PluginPreferences {
    type Plugin: Plugin;
    fn from_plugin_state(plugin: &Self::Plugin) -> Self;
    fn into_plugin_state(&self) -> Self::Plugin;
}
```

Reply:
> What does Preferences enable that the current plugin settings approach could not support / evolve into?

Plugin settings, as they are used currently, allow the developer to define the specific behavior they want. Stuff like "should the window exit when requested" or "how should we initialize wgpu".

These are distinct from user preferences. We don't want users tweaking them. We shouldn't mix developer settings and user-configurable preferences in the same type, that would be crazy!

Since we were just using the render plugin as an example, let's stick with it. The render plugin absolutely needs to expose preferences for visual fidelity if we want to target low end hardware. San Miguel runs at 10fps on my laptop, and I currently have no way to specify a different performance/fidelity trade off.

Implementing those options is a different problem, but it starts with having a place to put them that can be exposed to the user in a preferences screen or a file, and which isn't contaminated with settings not intended for user consumption.